File: modules/core-api/src/main/java/com/enonic/xp/content/page/region/ComponentType.java
package com.enonic.xp.content.page.region;

public abstract class ComponentType
{
    private final String shortName;

    private final Class componentClass;

    public ComponentType( final String shortName, final Class componentClass )
    {
        this.shortName = shortName;
        this.componentClass = componentClass;
    }

    public String toString()
    {
        return shortName;
    }

    public Class getComponentClass()
    {
        return componentClass;
    }
}

File: modules/admin/admin-ui/src/main/resources/web/admin/common/js/content/page/region/ComponentType.ts
module api.content.page.region {

    export class ComponentType {

        private static shortNameToInstance: {[shortName: string]: ComponentType} = {};

        private shortName: string;

        constructor(shortName: string) {
            ComponentType.shortNameToInstance[shortName] = this;
            this.shortName = shortName;
        }

        getShortName(): string {
            return this.shortName;
        }

        newComponentBuilder(): ComponentBuilder<Component> {
            throw new Error('Must be implemented by inheritors');
        }

        static byShortName(shortName: string): ComponentType {
            return ComponentType.shortNameToInstance[shortName];
        }

        getDefaultName(): ComponentName {
            return new ComponentName(api.util.StringHelper.capitalize(api.util.StringHelper.removeWhitespaces(this.shortName)));
        }
    }

}
